RA-196 failsafe snapshot support in EFSVSS

kevinhoffman opened the pull request:
This is required modifications to efsvss to support RA-196 functionality in the agent. Most of the logic will be in a companion PR within the agent.

This adds the following to efsvss:

`-wsml` command lists writer status in a format that is easier to parse. Also, this command no longer queries detailed writer metadata and only queries for status, so it is much faster and less resource intensive on the system.

`-snaplevel=x` option allows you to set a custom "snapshot level" value that will be remembered in the SETVAR script file that gets created. The snapshot level is user-defined and does not change any behavior of taking the VSS snapshot.

kevinhoffman committed:
RA-196: Add -wsml and -snaplevel to efsvss

-wsml command lists writer status in a format that is easier to parse. Also, this command no longer queries detailed writer metadata and only queries for status, so it is much faster and less resource intensive on the system.

-snaplevel=x option allows you to set a custom "snapshot level" value that will be remembered in the SETVAR script file that gets created. The snapshot level is user-defined and does not change any behavior of taking the VSS snapshot.

diazfran commented:
This is not necessary:
`Allow all builds to happen with Visual Studio 2015`

It won't compile for me, out of the box, if I try to build it in Visual Studio 2017 now.

I have reverted that commit from this branch in my system and it works without changing anything. It works in Visual Studio 2017 and in Visual Studio 2015.

Richard is uploading a BoM in Confluence today to make sure everyone knows what needs to be used and what features need to be enabled when you install Visual Studio.

kevinhoffman committed:
RA-196: snapshot creation and logging enhancements

- When gathering writer status after a VSS backup operation, always enumerate all writers before propagating the failure of writer(s).

- If DoSnapshotSet succeeds, but a writer has failed, make sure to mark VSS backup as failed.

- When comparing GUIDs in exclusion rules, allow the string GUID to still match even if it is not surrounded by curly brackets.

- Do not exclude the Shadow Copy Optimization Writer just because it has no components.

- Enhanced logging about excluded writers and writer status after a snapshot has been taken.

- New option -iwgf to ignore failures gathering metadata about individual writers.

kevinhoffman commented:
@diazfran - I'm rebasing this branch to remove the undesired VS2015 changes, so this can be merged without having to resolve that particular commit. Then I'll rebase my own personal branch with those changes so I can continue to build, but without it being in the main branch for this.
